transaction/core/src/membership_proofs/mod.rs

```diff
@@ -66,96 +199,29 @@ pub fn is_membership_proof_valid(
     known_root_hash: &[u8; 32],
 ) -> Result<bool, Error> {
     if proof.index > proof.highest_index {
-        return Ok(false);
+        return Err(Error::HighestIndexMismatch);
     }
-    // All Ranges contained in the proof must be valid. An invalid Range could be created
-    // by deserializing invalid bytes.
-    if proof.elements.iter().any(|e| e.range.from > e.range.to) {
-        return Err(Error::RangeError(RangeError {}));
-    }
-
-    // * The proof must contain the correct leaf hash for the specified index.
-    let leaf = Range::new(proof.index, proof.index)
-        .expect("A Range containing a single value is always well-formed.");
 
-    let range_to_hash: HashMap<Range, [u8; 32]> = proof
+    // The first element must be the leaf hash corresponding to the proof index
+    let first = proof
         .elements
-        .iter()
-        .map(|element| (element.range, *element.hash.as_ref()))
-        .collect();
-
-    if let Some(leaf_hash) = range_to_hash.get(&leaf) {
-        if *leaf_hash != hash_leaf(tx_out) {
-            // Proof contains incorrect leaf hash.
-            return Err(Error::IncorrectLeafHash(leaf.from));
-        }
-    } else {
-        // Proof does not contain a leaf hash for `tx_out`.
-        return Err(Error::MissingLeafHash(leaf.from));
-    }
-
-    // * The root hash of the proof must match the known root_hash.
-    let mut ranges: Vec<&Range> = range_to_hash.keys().collect();
-    ranges.sort();
-
-    let root_range = ranges.last().expect("`ranges` should be non-empty.");
-    let root_hash = range_to_hash
-        .get(root_range)
-        .expect("`root_range` should be a key.");
-
-    if *root_hash != *known_root_hash {
-        // Incorrect root hash.
-        return Ok(false);
+        .first()
+        .ok_or_else(|| Error::MissingLeafHash(proof.index))?;
+    if first.range.from != proof.index || first.range.to != proof.index {
+        return Err(Error::MissingLeafHash(proof.index));
     }
-
-    if proof.highest_index > root_range.to {
-        return Ok(false);
+    // The first element hash must match the tx_out
+    if !bool::from(first.hash.as_ref().ct_eq(&hash_leaf(tx_out))) {
+        return Err(Error::IncorrectLeafHash(first.range.from));
     }
 
-    // * All internal node's hashes between leaf and root must be recomputable from their children's hashes.
```

Comment:
after this change, the proof doesn't carry all the hashes of the internal nodes, those values are implied by the other values, so this check goes away, and we only check the final hash versus the known root hash